Add notes for running on bare-metal

fatal-exception opened the pull request.

fatal-exception committed:
[MR] Add bare-metal readme about vm_max_map_count sysctl

See https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html
and https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-configuration-memory.html#mlockall.
Elasticsearch requires the vm_max_map_count to be high or it will refuse
to start. Both the following options were rejected:
  - An annotation to set sysctls would require flag --experimental-allowed-unsafe-sysctls="vm.max_map_count"
  to be set on kubelets. Kops doesn't allow this, so would require kubespray to work
  differently from kops. (see https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/.)
  - initContainer - would require privileges so not that secure.

Instead, we just need to ensure that any nodes running elasticsearch
have vm.max_map_count set to 262144 in sysctl. Kops does this for us, so
on baremetal we just need to document it.

alanplatt commented on kubectl/instrumentation/elasticsearch/elasticsearch-data-statefulset.yaml:
```diff
@@ -49,6 +49,15 @@ spec:
+        imagePullPolicy: "Always"
+        command: ["sysctl", "-w", "vm.max_map_count=262144"]
+        securityContext:
+          privileged: true
```
@cherwin mentioned this in a thread https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/

Have you looked at that? Running privileged containers is best avoided

fatal-exception replied:
Thanks - re-working this now to add an annotation to the pods, and a value to the kubelet to allow sysctl changes

fatal-exception commented:
are you happy to merge now @alanplatt  ?

fatal-exception committed:
Add fix to alertmanager-sample so that alertmanager starts

The previous sample file had non-matching receiver names, so
alertmanager was in a crashing backoff loop

alanplatt commented on README.md:
```diff
@@ -1,2 +1,17 @@
+          configMap:
+            name: elasticsearch-startup-1
+        - name: elasticsearch-data
+          emptyDir: {}
```
This is interesting and I think it warrants further discussion. 

So before, we where using persistent storage to provide and extra layer of protection to an already resilient cluster system which seems sensible. Recovering from a 'hardware' loss is usually an expensive event for a cluster.  If we change this to use local storage we lose this extra level and change the behaviour of our cluster but in theory the cluster should still be fine and we can take the hit.

However, I think we should consider setting DefaultStorageClass for the cluster on physical hardware so that we can mimic a kops cluster as closely as possible. That way when an engineer is dealing with the different k8s clusters there are less differences? Plus we also keep our extra level of protection. Thoughts?

alanplatt replied:
@tommyvn @cherwin

fatal-exception replied:
sure - i was adding to this README to help anyone who builds a microdc on bare-metal in the meantime, to save them a bit of time troubleshooting.

We could look at rook.io to provide a platform-agnostic default storage class?
